---
title: "RNAseq Analysis Example"
output:
  html_document:
    keep_md: yes
---

This is an introduction to RNAseq analysis for use at Software Carpentry bootcamps that have covered novice R. It involves reading in some count data from an RNAseq experiment, exploring the data using base R functions and then analysis with the package DESeq2.

# Install required CRAN packages

First, install some packages that you'll use.

```{r install_packages, eval=FALSE}
install.packages("gplots")
install.packages("ggplot2")
install.packages("calibrate")
```

# Introduction and data import

The analysis of an RNAseq experiment begins with sequencing reads. These then need to be aligned to a reference genome or transcriptome. There are many different alignment tools available, but the process of alignment is both computationally intensive and time-consuming, so we won't cover it today. Once reads are aligned, the number of reads mapped to each gene can be counted. Again, there are several ways of doing this. The best way to find out about the tools that are available and suitable for your research is to look for recent review papers that compare the different tools.

The data for this tutorial comes from a PLOS ONE paper, [Genome-Wide Transcriptional Profiling of Skin and Dorsal Root Ganglia after Ultraviolet-B-Induced Inflammation](http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0093338)[1], and the raw data can be downloaded from the [Gene Expression Omnibus database (GEO)](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE54413). 

This data has already been downloaded and aligned to the human genome. The command line tool [featureCounts](http://bioinf.wehi.edu.au/featureCounts/) was used to count reads mapped to human genes from the Ensembl annotation (available for download [here](http://www.ensembl.org/info/data/ftp/index.html)). 

The output from this tool is provided in the `counts.txt` file. Have a look at this file in the shell, using `head`.

Import the data into R as a `data.frame` and examine it again. You can set the arguments of `read.table` to import the first row as a header giving the column names, and the first column as row names. 

```{r data_input}
# Filename with output from featureCounts
countFile <- "data/counts.txt"
# Read in the data
countData <- read.table(countFile, header=TRUE, row.names=1)
head(countData)
colnames(countData)
class(countData)
```

The data.frame contains information about genes (one gene per row) with the gene positions in the first five columns and then information about the number of reads aligning to the gene in each experimental sample. There are three replicates for control (column names starting with "ctl") and three for samples treated with ultraviolet-B light (starting "uvb"). We don't need the information on gene position for this analysis, just the counts for each gene and sample, so we can remove it from the data frame.

```{r}
# Remove first five columns (chr, start, end, strand, length)
countData <- countData[ ,-(1:5)]
head(countData)
colnames(countData)
```

We can rename the columns to something shorter and a bit more readable.

```{r eval=FALSE}
# Manually
c("ctl1", "ctl2", "ctl3", "uvb1", "uvb2", "uvb3")
# Using paste
?paste
paste("ctl", 1:3)
paste("ctl", 1:3, sep="")
?paste0
paste0("ctl", 1:3)
c(paste0("ctl", 1:3), paste0("uvb", 1:3))
```

An easier way to do this, especially for files with many columns, is to use the `gsub` command to strip out the extra information. This is also more robust to introduced errors, for example if the column order changes at some point in the future or you add additional replicates.

```{r rename_cols}
# Using gsub -- robust
?gsub
gsub(pattern=".fastq_tophat.accepted_hits.bam", replacement="", x=colnames(countData))
colnames(countData) <- gsub(pattern=".fastq_tophat.accepted_hits.bam", replacement="", x=colnames(countData))
head(countData)
```

## Exercise 1
Find the gene with the highest expression in any sample -- remember, each row is a gene. Extract the expression data for this gene for all samples. In which sample does it have the highest expression? 

What is the function of the gene? Can you suggest why this is the top expressed gene?

Hint 1: use the `apply` function from the introductory R lessons.

Hint 2: try `?which.max`.

```{r exercise_1, echo=FALSE, include=FALSE}
# use apply to get the maximum of each row, then get the maximum of these
max(apply(countData, 1, max)) # max expression is 7013
topGene <- which.max(apply(countData, 1, max)) # gene is EEF1A1P9
countData[topGene, ] # get other sample data - max is in uvb1

# this is a pseudogene - maybe an artefact of only aligning reads to a single chromosome?
```

# Data investigation using base R

We can investigate this data a bit more using some of the basic R functions before going on to use more sophisticated analysis tools.

First make a copy of the data, because we'll need it later. We will work on the copy. We will calculate the mean for each gene for each condition and plot them.

```{r get_means}
countData2 <- countData #make a copy

# get Control columns
colnames(countData2)
?grep #grep searches for matches to a pattern 
grep("ctl", colnames(countData2))
ctlCols <- grep("ctl", colnames(countData2))
head(countData2[,ctlCols])
head(apply(countData2[, ctlCols], 1, mean))
# here we'll use rowMeans instead, it's a convenient shortcut, and also faster!
countData2$ctlMean <- rowMeans(countData2[, ctlCols])

# same for uvb
uvbCols <- grep("uvb", colnames(countData2))
countData2$uvbMean <- rowMeans(countData2[, uvbCols])
```

Plot the mean expression of each gene in control against the UVB sample mean. Are there any outliers?

```{r plot_means}
plot(countData2$ctlMean, countData2$uvbMean)
```

```{r ggplot_means}
library("ggplot2")
ggplot(countData2, aes(x=ctlMean, y=uvbMean)) + geom_point()
```

## Exercise 2
How could you make this plot more informative and look more professional? 

Hint: try using a log scale. You can also changing colours, transparencies, sizes, or shapes of points. 

`?par` will give you information on lots of graphical parameters that can be set. Help for ggplot2 can be found [here](http://docs.ggplot2.org/current/).

```{r exercise2_1, echo=FALSE, include=FALSE}
plot(countData2$ctlMean, countData2$uvbMean, log="xy", pch=16)
```

```{r exercise2_2, echo=FALSE, include=FALSE}
ggplot(countData2, aes(x=ctlMean, y=uvbMean)) + geom_point() + scale_x_log10() + scale_y_log10() + theme_bw()
```

There are many more options you can use to alter the appearance of these plots.

# Find candidate differentially expressed genes

We can find candidate differentially expressed genes by looking for genes with a large change between control and UVB samples. A common threshold used is log2 fold change more than 2 or less than -2. We will calculate log2 fold change for all the genes and colour the genes with log2 fold change of more than 2 or less than -2 on the plot.

First, check for genes with a mean expression of 0. Putting zeroes into the log2 fold change calculation will produce NAs, so we might want to remove these genes. Note: this is for mathematical reasons, although different software may produce different results when you try to do `log2(0)`.

`TRUE` and `FALSE` can also be represented as 1 and 0. This is useful for getting the total number of observations for which a condition is true. 

```{r true_false}
TRUE == 0
TRUE == 1
FALSE == 0
```

This can be applied to testing whether genes have a mean expression of more than zero.

```{r count_expressed}
head(countData2$ctlMean)
head(countData2$ctlMean > 0)
head(as.numeric(countData2$ctlMean > 0))
```

When we call `sum(countData2$ctlMean > 0)`, we're really asking, "how many genes have a mean above 0 in the control group?"

```{r remove_unexpressed}
# discuss: why to remove zeroes (NAs produced)
sum(countData2$ctlMean > 0)
sum(countData2$uvbMean > 0)

nrow(countData2)
countData2 <- subset(countData2, (countData2$ctlMean > 0 | countData2$uvbMean > 0))
# explain: | operator meaning OR in this context?
nrow(countData2)
```

```{r log2FC}
# explain: what is fold change? why do we use log2 to quantify the fold change?
# discuss: Inf / -Inf may be produced in some cases. Concept of adding pseudocounts.
countData2$log2FC <- log2(countData2$uvbMean / countData2$ctlMean)
# again, reinforce that summing a logical vector gives you the number of 
# occurences of TRUE.
sum(countData2$log2FC > 2)
sum(countData2$log2FC < -2)
```

Make a new column to store this information in.

```{r outliers}
countData2$outlier <- FALSE
countData2$outlier[countData2$log2FC > 2] <- TRUE
countData2$outlier[countData2$log2FC < -2] <- TRUE
```

```{r plot_outliers}
plot(countData2$ctlMean, countData2$uvbMean, log="xy", pch=16)
points(countData2$ctlMean[countData2$outlier==TRUE], countData2$uvbMean[countData2$outlier==TRUE], col="red", pch=16)
```

```{r ggplot_outliers}
ggplot(countData2, aes(x=ctlMean, y=uvbMean, colour=outlier)) + geom_point() + scale_x_log10() + scale_y_log10() + theme_bw()
```

What do you notice about the positions of the outliers on these plots? How would you interpret this?

# DESeq2 analysis

DESeq2 is an R package for analysis of RNAseq data. It is available from [Bioconductor](http://www.bioconductor.org/). Bioconductor is a project to provide tools for analysing high-throughput genomic data including RNA-seq, ChIP-seq and arrays. You can explore Bioconductor packages [here](http://www.bioconductor.org/packages/release/BiocViews.html#___Software). 

```{r install_deseq2, eval=FALSE}
# install and have a break to check everyone is up to date?
# explain bioconductor?
source("http://bioconductor.org/biocLite.R")
biocLite("DESeq2")
```

```{r load_deseq2}
library("DESeq2")
citation("DESeq2")
```

It requires the count data to be in matrix form, and an additional dataframe describing the structure of the experiment.

```{r convert_to_matrix}
# countData is currently a data.frame, but DESeq2 expects its input to be in 
# matrix format, so we will convert our countData to a matrix.
class(countData)
countData <- as.matrix(countData)
class(countData)
head(countData)

# construct colData dataframe
# three replicates of control and UVB.
colData <- data.frame(condition=c(rep("ctl", 3), rep("uvb",3)), row.names=colnames(countData))
```

DESeq works on a particular type of object called a DESeqDataSet.

```{r make_deseqdataset}
# introduce how DESeq2 works - type of object it works on etc
# instantiate the DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData=countData, colData=colData, design=~condition)
dds
```

Run the DESeq pipeline on this object. [Describe pipeline steps?]
Get results and have a look at them

```{r run_deseq}
dds <- DESeq(dds)

# Get differential expression results
res <- results(dds)
head(res)
table(res$padj<0.05)
# Order by adjusted p-value
res <- res[order(res$padj), ]
head(res)
```

Combine DEseq results with the original counts data. Write significant results to a file.

```{r write_results}
resData <- merge(as.data.frame(res), as.data.frame(counts(dds, normalized=TRUE)), by="row.names", sort=FALSE)
head(resData)
names(resData)[1] <- "GeneID"
head(resData)

sig <- subset(resData, padj<0.05)
dir.create("results")
write.table(sig, file="results/sig.txt", sep="\t") # tab delim data
```

You can open this file in Excel or any text editor (try it now).

# Data Visualization

We can also do some exploratory plotting of the data.

```{r plot_dispersion}
plotDispEsts(dds, main="Dispersion plot")
```

```{r plot_heatmaps}
# Regularized log transformation for clustering/heatmaps, etc
rld <- rlogTransformation(dds)
plotPCA(rld)

# Sample distance heatmap
head(assay(rld))
assay(rld)[1:5,1:5]
t(assay(rld))[1:5,1:5]
dist(t(assay(rld)))
as.matrix(dist(t(assay(rld))))
sampleDists <- as.matrix(dist(t(assay(rld))))
heatmap(sampleDists)
# better heatmap with gplots
library("gplots")
heatmap.2(sampleDists)
heatmap.2(sampleDists, col=colorpanel(64, "steelblue", "white"), key=FALSE, trace="none")
heatmap.2(sampleDists, col=colorpanel(64, "black", "white"), key=FALSE, trace="none")
heatmap.2(sampleDists, col=colorpanel(64, "red", "black", "green"), key=FALSE, trace="none")
heatmap.2(sampleDists, col=colorpanel(64, "red", "white", "blue"), key=FALSE, trace="none")
```

```{r plot_pval_hist}
# Examine plot of p-values
hist(res$pvalue, breaks=50, col="grey")
```


```{r MA_plot}
# These are the plots that are most recognisable from papers
# MA Plot
par(pch=16)
with(res, plot(baseMean, log2FoldChange, pch=20, cex=.5, log="x"))
with(subset(res, padj<.05), points(baseMean, log2FoldChange, col="red", pch=16))
library("calibrate")
?textxy
res$Gene <- rownames(res)
with(subset(res, padj<.05), textxy(baseMean, log2FoldChange, labs=Gene, cex=1, col=2))
```

```{r volcano_plot}
# Volcano plot
# Set point character
par(pch=16)
with(res, plot(log2FoldChange, -log10(pvalue), main="Volcano plot"))
with(subset(res, padj<.05 ), points(log2FoldChange, -log10(pvalue), col="red"))
with(subset(res, abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), col="orange"))
with(subset(res, padj<.05 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), col="green"))
# Add legend
legend("topleft", legend=c("FDR<0.05", "|LFC|>1", "both"), pch=16, col=c("red","orange","green"))
# Label points
with(subset(res, padj<.05 & abs(log2FoldChange)>1), textxy(log2FoldChange, -log10(pvalue), labs=Gene, cex=1))
```

# References

1. Dawes JM, Antunes-Martins A, Perkins JR, Paterson KJ, Sisignano M, et al. (2014) Genome-Wide Transcriptional Profiling of Skin and Dorsal Root Ganglia after Ultraviolet-B-Induced Inflammation. PLoS ONE 9(4): e93338. doi: 10.1371/journal.pone.0093338 